### PR TITLE
[Web] Support web indexDB cache for larger model storage

### PR DESCRIPTION
**Pull request by DiegoCao**
Support IndexDB Cache, including the following:

- Implement ArtifactIndexedDBCache to cache model with IndexedDB.
- Add storetype to addToCache() and fetchWithCache() such that users need to specify either "json" or "arraybuffer" when using the APIs, instead of calling .json() or .arraybuffer() on their own outside of the function call.
- Move all cache-related code to artifact_cache.ts.
- Remove cacheOnly for callback in fetchNDArrayCache(), use loading instead (since we separated download and loading).

One change to use the IndexDBCache is, as there's no json() response when calling fetchCacheData, thinking about removing the .json() when calling fetchCacheData. @CharlieFRuan

### CODE REVIEW

**Review by CharlieFRuan — COMMENTED** [review]
Thank you! Tested with WebLLM and it is able to download and retrieve with IndexedDB. There are some items below we should address before merging it.

File: web/src/runtime.ts [hunk 1]
```diff
@@ -1052,6 +1062,191 @@ export class ArtifactCache implements ArtifactCacheTemplate {
...
+      };
+      request.onerror = (event) => {
+        reject((event.target as IDBRequest).error);
+      };
+    });
+  }
+
+  constructor(dbName: string){
+    this.dbName = dbName;
+  }
```
[the remark this answers is not in the document]
Comment (DiegoCao) [reply web/src/runtime.ts]: Resolved

File: web/src/index.ts [hunk 2]
```diff
@@ -22,7 +22,7 @@ export {
   PackedFunc, Module, NDArray,
   TVMArray, TVMObject, VirtualMachine,
   InitProgressCallback, InitProgressReport,
-  ArtifactCache, Instance, instantiate, hasNDArrayInCache, deleteNDArrayCache
+  ArtifactCache, ArtifactIndexDBCache, Instance, instantiate, hasNDArrayInCache, deleteNDArrayCache
```
[the remark this answers is not in the document]
Comment (DiegoCao) [reply web/src/index.ts]: Resolved

File: web/src/runtime.ts [hunk 3]
```diff
@@ -1052,6 +1062,191 @@ export class ArtifactCache implements ArtifactCacheTemplate {
...
+        return await response.json();
+      } else {
+        return response;
+      }
+    } catch (error) {
+      console.error("There was a problem fetching the data:", error);
+    }
+  }
+
+  async hasAllKeys(keys: string[]) :Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (DiegoCao) [reply web/src/runtime.ts:1239]: The issue is fixed

File: web/src/runtime.ts [hunk 4]
```diff
@@ -1052,6 +1062,191 @@ export class ArtifactCache implements ArtifactCacheTemplate {
...
+    let response: Response;
+    try {
+      response = await fetch(url);
+      if (!response.ok) {
+        throw new Error('Network response was not ok');
+      }
+      const response_copy = response.clone();
+      await this.addToIndexDB(url, response_copy, storetype);
+      if (storetype.toLowerCase() === "arraybuffer"){
+        return await response.arrayBuffer();
```
[the remark this answers is not in the document]
Comment (DiegoCao) [reply web/src/runtime.ts:1228]: I modified the code to make sure the two cache behaves the same in usage.

**Comment by CharlieFRuan** [comment]
Thank you for the updates! I will review soon today/tomorrow.

**Comment by MasterJH5574** [comment]
@tvm-bot rerun

**Review by tqchen — APPROVED** [review]

**Pull request merged**